Source/WebCore/dom/SpatialImageControls.cpp

```diff
@@ -0,0 +1,219 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SpatialImageControls.h"
+
+#include "ElementInlines.h"
+#include "ElementRareData.h"
+#include "Event.h"
+#include "EventLoop.h"
+#include "HTMLButtonElement.h"
+#include "HTMLDivElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLNames.h"
+#include "HTMLStyleElement.h"
+#include "MouseEvent.h"
+#include "RenderImage.h"
+#include "ShadowRoot.h"
+#include "TreeScopeInlines.h"
+#include "UserAgentStyleSheets.h"
+#include <wtf/text/AtomString.h>
+
+namespace WebCore {
+namespace SpatialImageControls {
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+
+static const AtomString& spatialImageControlsElementIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls"_s);
+    return identifier;
+}
+
+static const AtomString& spatialImageControlsButtonIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls-button"_s);
+    return identifier;
+}
+
+bool hasSpatialImageControls(const HTMLElement& element)
+{
+    RefPtr shadowRoot = element.shadowRoot();
+    if (!shadowRoot || shadowRoot->mode() != ShadowRootMode::UserAgent)
+        return false;
+
+    return shadowRoot->hasElementWithId(spatialImageControlsElementIdentifier());
+}
+
+static RefPtr<HTMLElement> spatialImageControlsHost(const Node& node)
+{
+    RefPtr host = dynamicDowncast<HTMLElement>(node.shadowHost());
+    if (!host)
+        return nullptr;
+
+    return hasSpatialImageControls(*host) ? host : nullptr;
+}
+
+bool isSpatialImageControlsButtonElement(const Element& element)
+{
+    return spatialImageControlsHost(element) && element.getIdAttribute() == spatialImageControlsButtonIdentifier();
+}
+
+bool shouldHaveSpatialControls(HTMLImageElement& element)
+{
+    if (!element.document().settings().spatialImageControlsEnabled())
+        return false;
+
+    bool hasSpatialcontrolsAttribute = element.hasAttributeWithoutSynchronization(HTMLNames::controlsAttr);
+
+    auto* cachedImage = element.cachedImage();
+    if (!cachedImage)
+        return false;
```

Comment:
This isn't right. cached image doesn't get set immediately when src attribute, etc... changes. We need to detect when an image gets loaded via ImageLoader. Maybe we wanna add a code to do this in `ImageLoader::didUpdateCachedImage`?

Reply:
Is cachedImage not loaded by the time `HTMLImageElement::didAttachRenderers()` runs? That's what kicks off this function call initially

Reply:
How does that work? You're calling `updateSpatialImageControls` in just `HTMLImageElement::attributeChanged`.

Reply:
Sorry misspoke earlier and didn't realize you already commented before I deleted my comment.

Refactored things a bit so `updateSpatialImageControls` gets called in exactly two places: `HTMLImageElement::attributeChanged` and `ImageLoader::notifyFinished` (the cached image will always be fully loaded by the latter). The former is just to catch if the `controls` attribute is added or removed by javascript after the image has already loaded. In the event that the image hasn't yet loaded by the time a JS call decides to add the `controls` attribute, `updateSpatialImageControls` should gracefully fail when called by `attributeChanged` but it will be called again by `notifyFinished`.